tests/TestCase/Collection/CollectionTest.php

```diff
@@ -2512,4 +2534,30 @@ protected function datePeriod($start, $end)
     {
         return new \DatePeriod(new \DateTime($start), new \DateInterval('P1D'), new \DateTime($end));
     }
+
+    /**
+     * Tests to ensure that collection classes extending ArrayIterator work as expected.
+     *
+     * @return void
+     */
+    public function testArrayIteratorExtend()
+    {
+        $iterator = new TestIterator(range(0, 10));
+
+        $this->assertTrue(method_exists($iterator, 'checkValues'));
+        $this->assertTrue($iterator->checkValues());
+
+        //We need to perform at least two collection operation to trigger the issue.
+        $newIterator = $iterator
+            ->filter(function ($item) {
+                return $item < 5;
+            })
+            ->reject(function ($item) {
+                return $item > 2;
+            });
+
+        $this->assertTrue(method_exists($newIterator, 'checkValues'), 'Our method has gone missing!');
+        $this->assertTrue($newIterator->checkValues());
```

Comment:
@lorenzo Is this a correct use of the Collection class? Personally, I don't think so. Because, for example, if the method of inner iterator is `getPrimeNumbers()`, people would expect that it returns array(1), as 2, 3, 5 and 7 have been already filtered by `filter()` and `reject()` calls. However it would returns array(1, 2, 3, 5, 7) actually, as Collection doesn't remove any elements in the original iterator. I thought that was why we blocked calling `ArrayIterator::count()` via the Collection class.

Refs  #11440

Reply:
It is another method, though. Not the results you get when iterating, so it could be argued that this is vivo the correct and at the same time not.

Reply:
If it's a feature I think it would be needed to documented in the Collection page of our cookbook explicitly. Also, I would have to revert my PR #11045, as `method_exists($collection, 'cuustomMethod')` no longer works. `IteretorIterator` seems to handle `method_exists()` specially if the `__call()` method is not overridden, and we would not be able to emulate the behavior by any PHP code.

Reply:
Let's not support that feature, then

Reply:
OK. Then I will send a patch to fix the problematic test case.